pkg/kv/kvserver/store.go

```diff
@@ -2166,48 +2162,105 @@ func (s *Store) startRangefeedUpdater(ctx context.Context) {
 			closedts.SideTransportCloseInterval.SetOnChange(&st.SV, confChanged)
 			RangeFeedRefreshInterval.SetOnChange(&st.SV, confChanged)
 
-			getInterval := func() time.Duration {
+			// refreshInterval defines the interval between closed timestamp
+			// notification delivery for a given range.
+			getRefreshInterval := func() time.Duration {
 				refresh := RangeFeedRefreshInterval.Get(&st.SV)
 				if refresh != 0 {
 					return refresh
 				}
 				return closedts.SideTransportCloseInterval.Get(&st.SV)
 			}
 
+			// smearInterval defines the interval between closed timestamp
+			// notification delivery to batches of ranges. See below for an
+			// explanation.
+			const smearInterval = 1 * time.Millisecond
+			smearLimiter := rate.NewLimiter(rate.Every(smearInterval), 1)
+
+			// sleepTimer is used to wait out the refreshInterval when a small
+			// number of rangefeeds are active.
+			sleepTimer := timeutil.NewTimer()
+			defer sleepTimer.Stop()
+
+			var replIDs roachpb.RangeIDSlice
+			ctx, cancel := s.stopper.WithCancelOnQuiesce(ctx)
+			defer cancel()
 			for {
-				interval := getInterval()
-				if interval > 0 {
-					timer.Reset(interval)
-				} else {
-					// Disable the side-transport.
-					timer.Stop()
-					timer = timeutil.NewTimer()
+				if ctx.Err() != nil {
+					return
 				}
-				select {
-				case <-timer.C:
-					timer.Read = true
-					s.rangefeedReplicas.Lock()
-					replIDs = replIDs[:0]
-					for replID := range s.rangefeedReplicas.m {
-						replIDs = append(replIDs, replID)
+				// Notify each replica with an active rangefeed to check for an
+				// updated closed timestamp.
+				s.rangefeedReplicas.Lock()
+				replIDs = replIDs[:0]
+				for replID := range s.rangefeedReplicas.m {
+					replIDs = append(replIDs, replID)
+				}
+				s.rangefeedReplicas.Unlock()
+				// Sort the range IDs so that we notify in the same order on each
+				// iteration. With the pacing below, this helps to ensure a consistent
+				// closed timestamp lag for each range. The closedts-rangefeed-updater
+				// refresh loop operates independently of incoming closed timestamp
+				// updates, so this does not favor some replicas over others.
+				sort.Sort(&replIDs)
```

Comment:
How strictly does the `RangeFeedRefreshInterval` (`"kv.rangefeed.closed_timestamp_refresh_interval"`) need to be honoured? The provided algorithm results in a consistent lag only when the range ID is in same "smear batch" between the runs (in particular, if the set of ranges with a rangefeed does not change).

A bad example: there are 1000 ranges during the first run, 5 at a time are updated each ms (for 200ms). Then 800 last rangefeeds are disabled, now there are 200 ranges. 1 range is processed every ms, so the 200th one will be updated at the end of 200ms. During the previous run though it was processed at the 40ms mark. So the lag is 360ms between the 2 updates for this range.

The good news is that the guaranteed lag is limited by `2 x RangeFeedRefreshInterval`. This was true before this change too: the range IDs would be iterated in random order, so similarly the lag could exceed the target interval. But the spread between the target and the actual interval used to be smaller because all the processing happened in bursty fashion at the beginning of the 200ms interval.

A "proper" solution would honour the last time a range was updated for deciding when to update it next. The problem with such solutions though is stickiness: in the example above with 1000 ranges reduced to 200 ranges, the 200 ranges would be destined to always be updated within the first 40ms out of 200ms. We wouldn't be able to easily rebalance them, although we could do **quicker** updates for some ranges so they shift to a different bucket.

Reply:
If we set `target = RangeFeedRefreshInterval` and we're ok to have the actual update interval between `[target, target*2)`, then the solution in this PR works well, and sorting makes it reach the target more often when the set of rangefeed ranges is semi-static.

Alternatively, we can set `target = RangeFeedRefreshInterval / 2`, then the given algorithm will never exceed the `RangeFeedRefreshInterval` which is good. But it comes at the cost of actually updating 2 times more frequently. Often we'll be 2 times quicker than configured.

Reply:
I think it's fine if the worst-case lag is 2x. The changefeed will most often coalesce these anyway, and only emit checkpoints e.g. every 30 seconds by default. End-users would only notice when setting this to something large like on the order of minutes. In those cases we could also end up tripping the stuck rangefeed setting and trigger restarts, but in some sense operators need to pick appropriate settings. If some users really don't want this, they can disable smearing.